scripts/installer/curl_install_pypi/install.py

```diff
@@ -0,0 +1,420 @@
+# at a specified file path that is the entry point into the promptflow.
+#
+# The latest versions of all promptflow command packages will be installed.
+#
+
+import os
+import sys
+import platform
+import stat
+import tarfile
+import tempfile
+import shutil
+import subprocess
+import hashlib
+from urllib.request import urlopen
+
+
+PF_DISPATCH_TEMPLATE = """#!/usr/bin/env bash
+export PF_INSTALLER=Script
+{install_dir}/bin/python -m promptflow._cli._pf.entry "$@"
+"""
+
+PFAZURE_DISPATCH_TEMPLATE = """#!/usr/bin/env bash
+{install_dir}/bin/python -m promptflow._cli._pf_azure.entry "$@"
+"""
+
+PFS_DISPATCH_TEMPLATE = """#!/usr/bin/env bash
+{install_dir}/bin/python -m promptflow._sdk._service.entry "$@"
+"""
+
+
+VIRTUALENV_VERSION = '16.7.11'
+VIRTUALENV_ARCHIVE = 'virtualenv-'+VIRTUALENV_VERSION+'.tar.gz'
+VIRTUALENV_DOWNLOAD_URL = 'https://pypi.python.org/packages/source/v/virtualenv/'+VIRTUALENV_ARCHIVE
+VIRTUALENV_ARCHIVE_SHA256 = '0f73ef551d889bf4b3241f1819aaf5f5c7e27259c1537255b1f63207552919b1'
+
+DEFAULT_INSTALL_DIR = os.path.expanduser(os.path.join('~', 'lib', 'promptflow'))
+DEFAULT_EXEC_DIR = os.path.expanduser(os.path.join('~', 'bin'))
+PF_EXECUTABLE_NAME = 'pf'
+PFAZURE_EXECUTABLE_NAME = 'pfazure'
+PFS_EXECUTABLE_NAME = 'pfs'
+
+
+USER_BASH_RC = os.path.expanduser(os.path.join('~', '.bashrc'))
+USER_BASH_PROFILE = os.path.expanduser(os.path.join('~', '.bash_profile'))
+
+
+class CLIInstallError(Exception):
+    pass
+
+
+def print_status(msg=''):
+    print('-- '+msg)
+
+
+def prompt_input(msg):
+    return input('\n===> '+msg)
+
+
+def prompt_input_with_default(msg, default):
+    if default:
+        return prompt_input("{} (leave blank to use '{}'): ".format(msg, default)) or default
+    else:
+        return prompt_input('{}: '.format(msg))
+
+
+def prompt_y_n(msg, default=None):
+    if default not in [None, 'y', 'n']:
+        raise ValueError("Valid values for default are 'y', 'n' or None")
+    y = 'Y' if default == 'y' else 'y'
+    n = 'N' if default == 'n' else 'n'
+    while True:
+        ans = prompt_input('{} ({}/{}): '.format(msg, y, n))
+        if ans.lower() == n.lower():
+            return False
+        if ans.lower() == y.lower():
+            return True
+        if default and not ans:
+            return default == y.lower()
+
+
+def exec_command(command_list, cwd=None, env=None):
+    print_status('Executing: '+str(command_list))
+    subprocess.check_call(command_list, cwd=cwd, env=env)
+
+
+def create_tmp_dir():
+    tmp_dir = tempfile.mkdtemp()
+    return tmp_dir
+
+
+def create_dir(dir):
+    if not os.path.isdir(dir):
+        print_status("Creating directory '{}'.".format(dir))
+        os.makedirs(dir)
+
+
+def is_valid_sha256sum(a_file, expected_sum):
+    sha256 = hashlib.sha256()
+    with open(a_file, 'rb') as f:
+        sha256.update(f.read())
+    computed_hash = sha256.hexdigest()
+    return expected_sum == computed_hash
+
+
+def create_virtualenv(tmp_dir, install_dir):
```

Comment:
what if venv already exists? do we always need to install?

Reply:
Actually virtual env will install in installation dir. If it's not empty, it will note user to delete it or choose another installation path.
![image](https://github.com/microsoft/promptflow/assets/26239730/96e3eca9-cd3d-4347-9d21-4023e4701fc1)

Reply:
Btw, if exec_dir also exists, it will overide pf/pfs/pfazure if user not changed exec_dir, so user will run latest pf when upgrade.

If user use a different exec_dir  to generate pf/pfs/pfazure, we will warn there're multiple pf in current machine
![image](https://github.com/microsoft/promptflow/assets/26239730/bad298d7-7a7e-4ae4-b38a-10ea67c6be4b)

Reply:
change to use python -m venv (>=python3.3)